### PR TITLE
feat: upgrade to AL2023 + EBS encryption by default

### PR DESCRIPTION
**Pull request by oycyc**
## what
* Upgrades the use of Amazon Linux 2 to Amazon Linux 2023
* Encrypt the EBS volume associated with the launch template by default
* 
Screenshot 2024-05-17 at 1 19 19 PM

* As a follow up, I will update the `terraform-aws-tailscale` module as well 

## references
* Addresses + closes https://github.com/masterpointio/terraform-aws-ssm-agent/issues/23
* Thanks @mheffner for bringing this up!

### CODE REVIEW

File: data.tf [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
-data "aws_ami" "amazon_linux_2" {
+# Most recent Amazon Linux 2023 AMI
+data "aws_ami" "amazon_linux_2023" {
   most_recent = true
   owners      = ["amazon"]
 
   filter {
     name   = "name"
-    values = ["amzn2-ami-hvm*"]
+    values = ["al2023-ami*"]
```
Comment (oycyc) [inline data.tf:11]: Their naming change of 2023 is interesting, before they just named it as AL2, now they're naming as AL2023

File: main.tf [hunk 2]
```diff
@@ -306,6 +306,13 @@ resource "aws_launch_template" "default" {
     create_before_destroy = true
   }
 
+  block_device_mappings {
+    device_name = "/dev/xvda"
```
Comment (oycyc) [inline main.tf:310]: This is a required field, I'm using the most standard default

File: data.tf [hunk 3]
```diff
@@ -1,18 +1,23 @@
...
     name   = "architecture"
     values = ["x86_64"]
   }
-}
+
+  filter {
+    name   = "virtualization-type"
+    values = ["hvm"]
+  }
+}
```
Comment (kevcube) [inline data.tf:23]: @oycyc can you enable settings in VSCode for auto-last-line

```json
  "files.insertFinalNewline": true,
  "files.trimFinalNewlines": true,
  "files.trimTrailingWhitespace": true,
```

File: main.tf [hunk 4]
```diff
@@ -271,7 +271,7 @@ DOC
 
 resource "aws_launch_template" "default" {
   name_prefix   = module.this.id
-  image_id      = length(var.ami) > 0 ? var.ami : data.aws_ami.amazon_linux_2.id
+  image_id      = length(var.ami) > 0 ? var.ami : data.aws_ami.amazon_linux_2023.id
```
Comment (kevcube) [inline main.tf:274]: We can do this cleaner with `coalesce()`

https://developer.hashicorp.com/terraform/language/functions/coalesce

Comment (oycyc) [reply main.tf]: Aha, thanks for that

File: data.tf (hunk 3 above)
Comment (oycyc) [reply data.tf]: I keep forgetting to do this for new installs 😓

Comment (oycyc) [reply data.tf]: Yeah... hmm it's there but the last empty line is just not showing up on the GitHub previews, same for the other files:
image
image

**Review by Gowiem — APPROVED** [review]
@oycyc this looks solid to me 👍

**Pull request merged**